### PR TITLE
Markdown support

### PR DESCRIPTION
**Pull request by f18m**
Adds support for --markdown-report option to generate output in Markdown syntax.

Note that
- I experimented with snippet rendering in markdown: I think the line numbers are important to have since they are mentioned in the diff coverage report. However from my investigations the only way to get the line numbers out of Pygments library is using either the HTML formatter or SVG formatter. Using the former the markdown looks OK, just without any colour.
- I didn't try yet to fix any testcase I broke (for unknown reasons now a couple of  HtmlReportGeneratorTest are now failing... despite the fact I didn't touch anything in html report generation)
- I didn't write yet any new testcase - but I'd like to have a feedback on the work so far

Please let me know what's your take on this work... thanks!

### CODE REVIEW

File: diff_cover/__init__.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # This file is imported in setup.py and cannot use any external dependencies
 
-VERSION = "4.0.1"
+VERSION = "4.0.2"
```
Comment (Bachmann1234) [inline diff_cover/__init__.py:3]: id prefer if you allowed me to handle this after merging. I like to handle all the packaging stuff together in a commit when im ready to release

File: diff_cover/report_generator.py [hunk 2]
```diff
@@ -337,19 +337,23 @@ def _src_path_stats(self, src_path):
...
+                snippet_list = Snippet.load_snippets(src_path, stats["violation_lines"])
+                snippets_html = [snippet.html() for snippet in snippet_list]
+                snippets_text = [snippet.terminal() for snippet in snippet_list]
             except OSError:
-                snippets = []
+                pass
         else:
-            snippets = []
+            snippets_html = []
+            snippets_text = []
```
Comment (Bachmann1234) [inline diff_cover/report_generator.py:351]: don't think you need this else any more

**Review by Bachmann1234 — COMMENTED** [review]
This is looking like its on the right track. My feedback is just nitpicks

I am also confused why that HTML test is failing after your changes. My only theory is that something is being consumed such that you cant loop over it twice. However, it is not clear to me where or why that would be.

**Comment by f18m** [comment]
Hi @Bachmann1234 ,
I think I finally found the reason why unit tests were failing: initially I bypassed the method Snippet.load_snippets_html() but I discovered that method was patched during unit testing.
Now I approached the problem by replacing load_snippets_html() with a new load_formatted_snippets() and patching during unit tests the latter method instead of the former.

Unit tests should be passing now.

File: diff_cover/__init__.py (hunk 1 above)
Comment (f18m) [reply diff_cover/__init__.py:3]: sure right I will revert this change - I used it just to make 100% I was using my local version of the upstream :)

File: diff_cover/report_generator.py (hunk 2 above)
Comment (f18m) [reply diff_cover/report_generator.py]: right - will remove

**Comment by Bachmann1234** [comment]
Looking good. Think this just needs some tests for the new functionality!

The build is still failing, but the tests are fine. Its sad because you need to run `black` to format your changes. Have you used it before? 

You should be able to install it and just run `black .` from the root of the project. 

Do those two and ill give the PR a final once over.

**Comment by f18m** [comment]
Hi @Bachmann1234 

> Looking good. Think this just needs some tests for the new functionality!
> 
> The build is still failing, but the tests are fine. Its sad because you need to run `black` to format your changes. Have you used it before?
> 
> You should be able to install it and just run `black .` from the root of the project.
> 
> Do those two and ill give the PR a final once over.

sure I will try to add some testcase to test_report_generator.py by emulating the testcases for the HTML formatter.

The major thing left out of this PR though is probably a decision: how to render in Markdown the code snippets.
A few options I see are:
a) create a markdown coverage report including code snippets that however contain HTML code without CSS which is not supported by most Markdown implementations. That means no coloured code anyway unless the colour CSS code is embedded straight into each <p> or <code> HTML tag using the "style" attribute.
Still we would get line numbers on the code "for free" at least.
Such solution does not look so good probably and it's not the "Markdown-way" likely.

b) try to use the TerminalFormatter of pygments library (https://pygments.org/docs/formatters/) that automatically adds the line numbers to the code - however it also places ANSI colour sequences in the output and these should be removed from the markdown output.

c) implement custom logic to add to the code snippets the line numbers, without any attempt to provide coloured code.

What do you think is best? Do you see any alternative?

Thanks

**Comment by f18m** [comment]
Answering to myself:

> What do you think is best? Do you see any alternative?

Adding another thought on this topic: using the "code-fenced blocks" of markdown (https://github.github.com/gfm/#fenced-code-blocks) means that the Markdown renderer implementation might try to apply syntax highlighting itself. Adding inside the fenced code block the line numbers of the code snippet  might however break the syntax highlighter (because the block wouldn't be anymore valid Python/Java/C++ code)

**Comment by Bachmann1234** [comment]
Ok this is looking pretty good but I forgot 1 thing. 

Can you update the readme? Should be small. For both tools (diff-cover, diff-quality) there is a line talking about the html/json reports. Just add the md one.

As for snippets with line numbers... 

Line number embedded into the snippet may be nice. Option C seems the most likely to work without fully breaking.

However it may be easier to just rely on the `code-fenced-blocks` solution you mentioned. Just dont embed the line numbers... something like

<snippet-start-line-num>-<snippet-end-line-num>
...code block...

might work nicely. 
example:

### lines 123-124
```
my_cool_var = 5
print(my_cool_var + 1)
```

**Comment by Bachmann1234** [comment]
However, end of the day if you prefer line numbers embeded it should not be ... _too_ hairy to remove those terminal codes

**Comment by Bachmann1234** [comment]
Readme looks great! When you think this is ready to merge just drop the WIP and ill give it one last scan, then merge/release!

**Comment by f18m** [comment]
> Readme looks great! When you think this is ready to merge just drop the WIP and ill give it one last scan, then merge/release!

great thanks - I think there's still some code to be written/tested for snippets in markdown syntax. I'm trying to add some unit test in test_snippet.py to test which of the different options (a, b or c) is the easiest / more effective solution... I agree that something like <snippet-start-line-num>-<snippet-end-line-num> just before each snippet would probably be very effective as solution.
The programmer doesn't need to have the exact lines pinned down probably - he just needs to have the context of which code is not being tested. At least in my case I'm using diff_cover for C/C++ code through gcov/lcov and lcov generates a nice HTML report where I can good to actually look in details at every single line of code and see how many times it was executed, branch coverage etc.

**Comment by f18m** [comment]
Hi @Bachmann1234 ,
I have completed the implementation following your suggestion to simply indicate before each fenced code block the lines of that snippet. I think the output is quite clear and easy to follow.
The new functions should be all covered by test cases. 
I have removed the WIP label... let me know if you think something else is needed.

**Comment by Bachmann1234** [comment]
This looks good to me. Ill merge this now and hopefully put out a release very soon

**Comment by Bachmann1234** [comment]
Thanks for the PR btw. This seems like a fun feature!

**Comment by Bachmann1234** [comment]
this is out! https://pypi.org/project/diff-cover/4.1.0/

**Pull request merged**